event_routing_backends/helpers.py

```diff
@@ -116,7 +116,8 @@ def get_block_id_from_event_referrer(event):
         str or None
     """
     try:
-        parsed = urlparse(event['context']['referer'])
+        referrer = event.get('referer', event['context']['referer'])
+        parsed = urlparse(referrer)
```

Comment:
Does the event sometimes have referer at the top level and sometimes under the context key? Was this necessary for removing the referer, or is it a bugfix that's coming along for the ride?

Reply:
yeah it appears event sometimes has `referer` at top level and sometimes under context. we had these statement
`get_block_id_from_event_referrer(self.event) or self.event['context']['referer']` earlier I have removed second part of it and made that change. However, It appears we were setting referrer to `object_id` if `get_block_id_from_event_referrer` returns `None`. Now that we cannot set referer to `object_id` what should we set `object_id` if `get_block_id_from_event_referrer` method return `None`?

Reply:
In what circumstances would `referer` be at the top level? I don't see any such cases in the OEP or in the test events.

It's unfortunate that we're parsing object ID from referer. Maybe we can change the events to add it more explicitly? But in the meantime, I think we'll have to assume that it's always available—there are a lot of things that could change about the source events and break processing, and I'm not sure there's much we could do about it.

Reply:
@timmc-edx  [LegacyFieldMappingProcessor](https://github.com/edx/edx-platform/blob/b7cd41f100dfcd6bf59ff8f0e7466badddd92ae5/common/djangoapps/track/shim.py#L19)  is responsible for moving referer from context to top level. Let me add test for case when we don't have object id in referer and referer is at top level.
I think if we are unable to find object ID from referer we should generate UUID and assign it as object id for that event. What you think?

Reply:
If there are cases where the referer might be in a different place, that should be represented in test JSON files.

I don't think I know enough about object ID. I guess my inclination would be to just fail if we can't find the block ID for an event that needs it.

Reply:
@timmc-edx it appears  LegacyFieldMappingProcessor is only used for `tracking_logs` backend so for `caliper` or `xAPI` we would not get `referrer` in top level event. I have updated PR to omit code which look for `referrer` at top level and also raising exception if object ID is not present in `referrer`.